src/export/digitalPlanning/model.ts

```diff
@@ -133,14 +143,60 @@ export class DigitalPlanning {
     };
   }
 
+  mapPassportToPreApplicationPayload(): PreApplicationPayload {
+    return {
+      data: {
+        user: {
+          role: this.passport.data?.["user.role"]?.[0],
+        },
+        applicant:
+          this.passport.data?.["user.role"]?.[0] === "applicant"
+            ? this.getApplicant()
+            : this.getApplicantWithAgent(),
+        property: this.getProperty(),
+        application: {
+          type: {
+            value: "preApp",
+            description: "Pre-application",
+          },
+          fee: {
+            payable: this.passport.data?.["application.fee.payable"] as number,
+            // Account for self-pay (has passport data) or invite to pay (has govUkPayment only)
+            ...((this.passport.data?.["application.fee.reference.govPay"] ||
+              this.govUkPayment) && {
+              reference: {
+                govPay:
+                  this.passport.data?.["application.fee.reference.govPay"]?.[
+                    "payment_id"
+                  ] || this.govUkPayment?.payment_id,
+              },
+            }),
+          },
+          declaration: this.getApplicationDeclaration(),
+          information: this.getPreAppInformation(),
+        },
+        proposal: {
+          description: this.passport.data?.["proposal.description"] as string,
+          boundary: this.getProposedBoundary(),
+        },
+      },
+      responses: this.getResponses(),
+      files: this.getFiles(),
+      metadata: this.getMetadata() as PlanXMetadata,
+    };
+  }
+
   validatePayload() {
     const ajv = addFormats.default(new Ajv.default({ allowUnionTypes: true }));
-    const validate = ajv.compile(jsonSchema);
+    const validate =
+      this.applicationType === "preApp"
+        ? ajv.compile(preApplicationJsonSchema)
+        : ajv.compile(applicationJsonSchema);
```

Comment:
One thought on the whole multiple classes thing I've just hit seeing some of this code. 

We can inherit based on classes so maybe there's a decent way forward here without a huge rewrite that has the majority of the functionality in a single abstract / base class, and then have per-app type classes that _just_ handle the differences. 

This might actually make things a whole lot simpler - highlighting the differences and maintaining the joint functionality.

Very happy to have a pass at this so we can compare and make a decision - worth exploring before going for a third schema 👍

Reply:
Super nice idea :bookmark: